File: 1157_online_majority_element_in_subarray/solution.php
<?php
// LeetCode 1157 - Online Majority Element In Subarray
// https://leetcode.com/problems/online-majority-element-in-subarray/

class MajorityChecker {
    private $arr;
    private $pos = [];

    /**
     * @param Integer[] $arr
     */
    function __construct($arr) {
        $this->arr = $arr;
        foreach ($arr as $i => $x) {
            $this->pos[$x][] = $i;
        }
    }

    /**
     * @param Integer $left
     * @param Integer $right
     * @param Integer $threshold
     * @return Integer
     */
    function query($left, $right, $threshold) {
        $candidate = 0;
        $count = 0;
        for ($i = $left; $i <= $right; $i++) {
            if ($count === 0) $candidate = $this->arr[$i];
            $count += $this->arr[$i] === $candidate ? 1 : -1;
        }
        $locs = $this->pos[$candidate] ?? [];
        $lo = 0; $hi = count($locs);
        while ($lo < $hi) {
            $mid = ($lo + $hi) >> 1;
            if ($locs[$mid] < $left) $lo = $mid + 1;
            else $hi = $mid;
        }
        $L = $lo;
        $lo = 0; $hi = count($locs);
        while ($lo < $hi) {
            $mid = ($lo + $hi) >> 1;
            if ($locs[$mid] <= $right) $lo = $mid + 1;
            else $hi = $mid;
        }
        return ($lo - $L) >= $threshold ? $candidate : -1;
    }
}

File: 1157_online_majority_element_in_subarray/Solution.scala
// LeetCode 1157 - Online Majority Element In Subarray
// https://leetcode.com/problems/online-majority-element-in-subarray/

class MajorityChecker(arr: Array[Int]) {
  private val pos = scala.collection.mutable.Map.empty[Int, scala.collection.mutable.ArrayBuffer[Int]]
  for (i <- arr.indices) {
    pos.getOrElseUpdate(arr(i), scala.collection.mutable.ArrayBuffer.empty) += i
  }

  def query(left: Int, right: Int, threshold: Int): Int = {
    var candidate = 0
    var count = 0
    for (i <- left to right) {
      if (count == 0) candidate = arr(i)
      count += (if (arr(i) == candidate) 1 else -1)
    }
    val locs = pos(candidate)
    def lower(x: Int): Int = {
      var lo = 0
      var hi = locs.length
      while (lo < hi) {
        val mid = (lo + hi) / 2
        if (locs(mid) < x) lo = mid + 1 else hi = mid
      }
      lo
    }
    def upper(x: Int): Int = {
      var lo = 0
      var hi = locs.length
      while (lo < hi) {
        val mid = (lo + hi) / 2
        if (locs(mid) <= x) lo = mid + 1 else hi = mid
      }
      lo
    }
    val freq = upper(right) - lower(left)
    if (freq >= threshold) candidate else -1
  }
}
